broadinstitute/single_cell_portal_core
# Source SECRET_KEY_BASE from vault (SCP-3155)

Currently, every time the portal is re-deployed, we generate a new value for `SECRET_KEY_BASE` at runtime and pass this to the Docker container.  This value is used as the [input value](https://api.rubyonrails.org/v5.2.1/classes/Rails/Application.html#method-i-secret_key_base) to the application's secret generator, for things such as secure cookies, and also as the salt for encrypting user's OAuth refresh tokens.  While this is very secure - the value is generated at runtime and never persisted anywhere in code/configuration - it requires us to clear all saved user tokens every time the portal re-deploys.  This leads to a degraded UX that causes all sessions to be invalidated, and causes many `422: InvalidAuthenticityToken` errors as the CSRF token in client browsers cannot be decoded anymore. 

This update replaces the runtime generation for `SECRET_KEY_BASE` and instead sources a static value from `vault`.  This will require a migration to clear all currently saved tokens as `SECRET_KEY_BASE` is changing, but this will be the final time it changes.  All subsequent releases/deployments should no longer sign out users or clear cached tokens (although tokens will still be deleted on sign out as per our Privacy Policy).

To test:
1. Pull this branch 
1. Run `rails_local_setup.rb` to pull all secrets from `vault`
1. Run `bin/rails db:migrate` to unset all stored user tokens
1. Boot server as normal, and sign in a user
1. Stop the server, and re-run `rails_local_setup.rb`
1. Restart the server, and refresh the page.  You should still be signed in, and there should not have been any `OpenSSL` errors when decrypting stored tokens in the database

This PR satisfies SCP-3155
